### ISSUE TITLE
Waves on water is broken

### ISSUE THREAD

**Issue by Miner34dev** [RigsOfRods/rigs-of-rods#3137]
**Hardware**: AMD Ryzen 3 5300U, integrated graphics
**Software**: Garuda Linux, latest RoR dev build, OpenGL render (of course)
**Settings**: [RoRcfg.txt](https://github.com/RigsOfRods/rigs-of-rods/files/14495442/RoRcfg.txt)

HydraX with waves on water glitches out badly (but only if you have a vehicle spawned, else it works as normal):
![Screenshot_20240305_132508](https://github.com/RigsOfRods/rigs-of-rods/assets/134534394/9a7cae40-5600-43d8-87aa-8a5a713ab785)
HydraX without waves on water somehow seems to have waves actually:
![Screenshot_20240305_133006](https://github.com/RigsOfRods/rigs-of-rods/assets/134534394/c026bef9-50ed-4e73-b7a7-402e6a81d8f5)
Reflection + refraction with waves on water doesn't have waves at all:
![Screenshot_20240305_132818](https://github.com/RigsOfRods/rigs-of-rods/assets/134534394/fec73a37-c804-4cb1-aeac-34e9383a5c36)

**Comment by ohlidalp** [comment]
Hello.

The 'waves on water' is somewhat arcane feature, I don't think anybody uses it because it hurts FPS. It works by creating a grid mesh and shaping it dynamically to simulate a wavy sea, both physically and visually.

Hydrax comes with fake caustics, that's what you're seeing.

**Comment by ohlidalp** [comment]
I found a crash scenario (Win10, Directx9, PSSM off, configuration=RelWithDebInfo) - OGRE crashes when rendering a frame:
![obrazek](https://github.com/RigsOfRods/rigs-of-rods/assets/491088/59a9310f-38b3-4345-ad3a-32c68ad21737)
Steps to reproduce:
1. In main menu, go to settings, set water=basic and Waves=on 
2. Load a map with water, like 'Small Island' (the one on screenshots above).
3. Slightly move the 'Wave height' slider in 'Top menubar/Settings' menu.
4. Go back to main menu, set water=Hydrax and resume game
5. Go back to main menu again -> crash.

I'd try under Debug but there's a showstopper - the terrain calculates lightmap on background thread, and you can't return to menu until that finished. It takes ~15sec on RelWithDebInfo. On Debug it would take at least 5min. I need to code an option to force-disable lightmap calc to test.

**Comment by ohlidalp** [comment]
I managed to reproduce the issue (Win10, Directx9, Visual Studio 2022, config=RelWithDebInfo, terrain=Small Island).
![obrazek](https://github.com/RigsOfRods/rigs-of-rods/assets/491088/866eb29a-6376-4c23-a14c-f3c6fec635cc)
Apparently it gets worse the more vehicles (or the more complex vehicle) is simulated.
![obrazek](https://github.com/RigsOfRods/rigs-of-rods/assets/491088/f2a71544-3975-47e2-bad1-6c88c0ee11e4)

**Comment by ohlidalp** [comment]
I couldn't figure anything out from the code, so I added GUI to fiddle with HydraX params...
![obrazek](https://github.com/RigsOfRods/rigs-of-rods/assets/491088/84579958-9843-4ce1-b56b-e64b3858573c)

anyway, when I figure out what causes this glitch, I'll try to replicate it as rain effect. It really looks like pouring water.

**Comment by cryham** [comment]
Yeah I [made a video](https://youtu.be/d-_uhZ1kybo) with this and other bugs some time ago.

**Comment by ohlidalp** [comment]
> And my guess was that you have some uninited vars for water.

You win. The glitch of Basic water was unitialized variable `Water::m_waves_height`. I have no progress on the Hydrax glitch though.

**Comment by cryham** [comment]
Cool.
Well since @tritonas00is working on the new water (btw is it on a branch?) , I also think it'd be best to finish that, and just forget Hydrax.

**Comment by ohlidalp** [comment]
I didn't find the issue yet, but I'm getting close ... I made a debug UI displaying min/max vertex height, so now I know the problem is in C++ code updating the water mesh.
![obrazek](https://github.com/RigsOfRods/rigs-of-rods/assets/491088/9643b73e-b2e1-4365-a01c-333a3a999974)
Yes, Tritonific water is super promising, but I haven't looked at it yet. I was just curious if this can be patched easily - from my experience on this project, quick and dirty solutions usually work out better that longer term projects.
